jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java

```diff
@@ -1,139 +1,168 @@
-            } else if( field.isRequired() ) {
-                throw new JGivenMissingRequiredScenarioStateException( field.getField() );
+                log.debug("Setting field {} to value {}", field.getField(), value);
+            } else if (field.isRequired()) {
+                throw new JGivenMissingRequiredScenarioStateException(field.getField());
             }
         }
     }
 
-    public <T> void injectValueByType( Class<T> clazz, T value ) {
-        state.updateValueByType( clazz, value );
+    public <T> void injectValueByType(Class<T> clazz, T value) {
+        state.updateValueByType(clazz, value);
     }
 
-    public <T> void injectValueByName( String name, T value ) {
-        state.updateValueByName( name, value );
+    public <T> void injectValueByName(String name, T value) {
+        state.updateValueByName(name, value);
     }
 
-    private void updateValue( ScenarioStateField field, Object value ) {
-        if( field.getResolution() == Resolution.NAME ) {
-            state.updateValueByName( field.getField().getName(), value );
+    private void updateValue(ScenarioStateField field, Object value) {
+        if (field.getResolution() == Resolution.NAME) {
+            state.updateValueByName(field.getField().getName(), value);
         } else {
-            state.updateValueByType( field.getField().getType(), value );
+            state.updateValueByType(field.getField().getType(), value);
         }
     }
 
-    private Object getValue( ScenarioStateField field ) {
-        if( field.getResolution() == Resolution.NAME ) {
-            return state.getValueByName( field.getField().getName() );
+    private Object getValue(ScenarioStateField field) {
+        if (field.getResolution() == Resolution.NAME) {
+            return state.getValueByName(field.getField().getName());
         }
 
-        return state.getValueByType( field.getField().getType() );
+        return state.getValueByType(field.getField().getType());
+    }
+
+    private void checkGuaranteedStatesAreInitialized(Object instance) {
+        for (Field field: FieldCache.get(instance.getClass())
+                .getFieldsWithAnnotation(ProvidedScenarioState.class, ScenarioState.class)) {
+            if (field.isAnnotationPresent(ProvidedScenarioState.class)) {
+                if (field.getAnnotation(ProvidedScenarioState.class).guaranteed()) {
+                    checkInitialized(instance, field);
+                }
+            }
+            if (field.isAnnotationPresent(ScenarioState.class)) {
+                if (field.getAnnotation(ScenarioState.class).guaranteed()) {
+                    checkInitialized(instance, field);
+                }
+            }
+        }
+    }
+
+    private void checkInitialized(Object instance, Field field) {
+        Object value = null;
+        try {
+            value = field.get(instance);
+        } catch (IllegalAccessException e) { }
```

Comment:
I wouldn't lump the Illegal access and the absent value excpetion together. They are two very different kinds of errors with very different reasons.